drivers: otp: add OTP driver subsystem

GseoC opened the pull request:
This P-R is a subset of #98149 and contains only the OTP subsystem and it's usage by the NVMEM subsystem.

Since the split of the P-R, I added some bits of documentation that I checked building the doc.

GseoC committed:
drivers: Introduce new OTP subsystem

Introduce a new OTP subsystem to be able to interact with One Time
Programmable(OTP) memory. For now, add basic read()/program() APIs.
Program() API is default disabled due to its sensitivity.

File drivers/otp.h is inspired by drivers/eeprom.h as the basic
features are similar.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
nvmem: Add OTP API support

OTP drivers are the interface to One Time Programmable memory, which
is NVMEM. Add the interface with the OTP API to the NVMEM one.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
tests: nvmem: Add OTP testcase

Add an overlay/config to test the NVMEM subsystem with an emulated
OTP driver.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
drivers: otp: add OTP memory emulation driver

In order to make some basic tests on the OTP API, add an OTP memory
emulator driver. It implements the .program() and .read() APIs.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC committed:
boards: native_sim: add OTP memory emulator node

Add an OTP memory emulator node to the board native_sim board device
tree file.

Additions of a node to the native sim device tree shifts the device
index, hence making the check_init_priorities test fail. Update the
index in the reference data of the test.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

aescolar commented on doc/hardware/peripherals/otp/api.rst:
```diff
@@ -0,0 +1,23 @@
+Overview
+********
+
+The OTP API provides means to provision and read with One Time Programmable(OTP)
```
minor suggestion:
```suggestion
The OTP API provides means to provision and read One Time Programmable(OTP)
```

aescolar reviewed (commented):
Just 2 minor readability suggestions.
+1 otherwise

GseoC commented on doc/hardware/peripherals/otp/index.rst:
```diff
@@ -0,0 +1,17 @@
+
+OTP memory devices handle memory that is expected to be permanent. This memory
+usually provisionned at manufacturing level but there are some use-cases
+that need to program OTP memory at a later lifecycle state (e.g: monotonic
```
I like the term **provisioned** for this kind of technology. It is widely used for it.
I'll apply the rest of your suggestion.

pdgendt commented on doc/hardware/peripherals/otp/api.rst:
```diff
@@ -0,0 +1,23 @@
+Overview
+********
+
+The OTP API provides means to provision and read One Time Programmable(OTP)
```
We have a doc directive command for abbreviations

```suggestion
The OTP API provides means to provision and read :abbr:`OTP (One Time Programmable)`
```

aescolar reviewed (dismissed):
+1

pdgendt commented on include/zephyr/drivers/otp.h:
```diff
@@ -0,0 +1,118 @@
+
+	if (len == 0U) {
+		return -EINVAL;
+	}
```
I'll voice again that I don't think this belongs here, but not blocking.

pdgendt reviewed (commented):
LGTM, one missing bit is a new maintainer area in the `MAINTAINERS.yml` file. But not sure how to proceed here.

GseoC committed:
MAINTAINERS: Add area for OTP

Create a section for the OTP subsystem. NVMEM is referenced in the test
category as the OTP subsystem is tested using the NVMEM API.

Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>

GseoC commented:
Added entry in MAINTAINERS file for the OTP subsystem

GseoC commented:
Fixed failed compliancy check because of uppercase sorting.

pdgendt commented:
@GseoC PTAL at the CI twister issues, they seem to be caused by this PR:

```shell
$ west build -p -b native_sim/native/64 tests/misc/check_init_priorities -T init.check_init_priorities
...
Mismatched otuput from /home/pdgendt/zephyrproject/zephyr/build/zephyr/check_init_priorities_output_initlevels.txt

expected:
APPLICATION
EARLY
POST_KERNEL
PRE_KERNEL_1
PRE_KERNEL_2
SMP
__init___device_dts_ord_31: init_fn_0(__device_dts_ord_31)
__init___device_dts_ord_32: init_fn_1(__device_dts_ord_32)
__init___device_dts_ord_33: NULL(__device_dts_ord_33)
__init___device_dts_ord_34: NULL(__device_dts_ord_34)
__init_posix_arch_console_init: posix_arch_console_init(NULL)
__init_sys_clock_driver_init: sys_clock_driver_init(NULL)

got:
APPLICATION
EARLY
POST_KERNEL
PRE_KERNEL_1
PRE_KERNEL_2
SMP
__init___device_dts_ord_32: init_fn_0(__device_dts_ord_32)
__init___device_dts_ord_33: init_fn_1(__device_dts_ord_33)
__init___device_dts_ord_34: NULL(__device_dts_ord_34)
__init___device_dts_ord_35: NULL(__device_dts_ord_35)
__init_posix_arch_console_init: posix_arch_console_init(NULL)
__init_sys_clock_driver_init: sys_clock_driver_init(NULL)
TEST FAILED
```

pdgendt commented:
> I fixed the expected output for now.

Strictly speaking, to not break bisectability, the fixup commit should be squashed with the commit that introduces the check failure.

GseoC commented:
> > I fixed the expected output for now.
> 
> Strictly speaking, to not break bisectability, the fixup commit should be squashed with the commit that introduces the check failure.

Squashed the test update commit into the dts one.

JarmouniA commented:
> I can downgrade myself to collaborator in the MAINTAINERS file as well if needed.

That would be appropriate in my opinion, taking into consideration the description of the Maintainer role.

https://docs.zephyrproject.org/latest/project/project_roles.html#maintainer

pdgendt commented:
Might need a rebase to fix the CI issues? Don't see the errors on `main` locally.

carlescufi reviewed (approved):
> Is it clear on the intent and scope of the API?

Yes

> Is it aligned with industry norms?

I searched and investigated quite thoroughly and, to me, OTP is the right name choice.

> Is it preferred over efuse (or maybe others)?

Yes, because according to most sources I have looked at, efuse is a  *type* of OTP, just like anti-fuse or others

GseoC commented:
> > Great news, thank you all.
> 
> Great work! Can you also create a PR with an entry in the release notes?

I'll have a look!
